source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst

```diff
@@ -0,0 +1,87 @@
+--------
+Plugins provide the ability to create custom widgets, layouts, data sources/types, and custom themes. Shuffleboard provides the following plugins:
+
+- NetworkTables Plugin: To view data published on the NetworkTables
+- Base Plugin: To display custom FRC data types in custom widgets
+- CameraServer Plugin: To view streams from the CameraServer
+
+Create a Custom Plugin
+----------------------
+In order to define a plugin, the plugin class must be a subclass of `edu.wpi.first.shuffleboard.api.Plugin <https://github.com/wpilibsuite/shuffleboard/blob/master/api/src/main/java/edu/wpi/first/shuffleboard/api/plugin/Plugin.java>`_ or one of its subclasses. An example of a plugin class would be as following.
+
+.. tabs::
+
+   .. code-tab:: java
+
+      import edu.wpi.first.shuffleboard.api.plugin.Description;
+      import edu.wpi.first.shuffleboard.api.plugin.Plugin;
+
+      @Description(group = "com.example", name = "MyPlugin", version = "1.2.3", summary = "An example plugin")
+      public class MyPlugin extends Plugin {
+
+      }
+
+Note the ``@Description`` annotation which is needed to tell the plugin loader the properties of the custom plugin class.
+Plugin classes are permitted to have a defualt constructor but it cannot take any arguments.
+
+Building plugin
+---------------
+Plugins require the usuage of the `Shuffleboard API Library <https://frcmaven.wpi.edu/artifactory/release/edu/wpi/first/shuffleboard/api/>`_. These dependencies can be resolved in the
+``build.gradle`` file or using maven. The dependencies would be as follows:
+
+For Gradle:
+
+.. code-block:: groovy
+
+   dependencies {
+      compileOnly("edu.wpi.first.shuffleboard", "api", "2020.+")
+      compileOnly("edu.wpi.first.shuffleboard.plugin", "networktables", "2020.+")
+   }
+
+Plugins are allowed to have dependencies on other plugins and libraries, however, they must be included correctly in the maven or gradle build file.
+When having Plugin dependencies to other plugins, it is good practise to define those dependencies so the plugin does not load when the dependencies do not load as well.
+This can be done using the ``@Requires`` annotation as shown below:
+
+.. code-block:: java
+
+   @Requires(group = "com.example", name = "Good Plugin", minVersion = "1.2.3")
+   @Requires(group = "edu.wpi.first.shuffleboard", "Base", minVersion = "1.0.0")
+   @Description(group = "com.example", name = "MyPlugin", version = "1.2.3", summary = "An example plugin")
+   public class MyPlugin extends Plugin {
+
+   }
+
+Deploying Plugin To Shuffleboard
+--------------------------------
+In order to load a plugin to the shuffleboard, you will need to generate a jar file of the plugin and put it in the ``~/Shuffleboard/plugins`` folder. This can be done automatically
+from gradle as noted:
+
+.. code-block::groovy
+
+   task deployWidget (type: Copy, group: "...", description: "...", dependsOn: "build") {
+      from "build/libs"
+      into "path/to/Shuffleboard/plugins"
```

Comment:
Typically `~/Shuffleboard/plugins`, though it might end up being `~/wpilib/Shuffleboard/plugins`